File: Ejercicios/proxy.py
from abc import ABC, abstractmethod


class Dron(ABC):
    @abstractmethod
    def enviar_paquete(self, paquete):
        pass


class DronReal(Dron):
    def __init__(self, id_dron):
        self.id_dron = id_dron

    def enviar_paquete(self, paquete):
        print(f"Dron {self.id_dron} enviando paquete: {paquete}")


class ProxyDron(Dron):
    def __init__(self, id_dron, usuario):
        self.id_dron = id_dron
        self.usuario = usuario
        self.dron_real = None
        self.usuarios_autorizados = ["admin", "operador1"]

    def enviar_paquete(self, paquete):
        if self.usuario in self.usuarios_autorizados:
            if self.dron_real is None:
                self.dron_real = DronReal(self.id_dron)
            self.dron_real.enviar_paquete(paquete)
        else:
            print(f"Acceso denegado al dron {self.id_dron} para el usuario: {self.usuario}")


dron1 = ProxyDron("DRN-001", "admin")
dron1.enviar_paquete("Paquete 1")

dron2 = ProxyDron("DRN-002", "invitado")
dron2.enviar_paquete("Paquete 2")

File: Ejercicios/Proxy.java
package Ejercicios;


interface Dron {
    void enviarPaquete(String paquete);
}


class DronReal implements Dron {
    private String idDron;

    public DronReal(String idDron) {
        this.idDron = idDron;
    }

    public void enviarPaquete(String paquete) {
        System.out.println("Dron " + idDron + " enviando paquete: " + paquete);
    }
}


class ProxyDron implements Dron {
    private String idDron;
    private String usuario;
    private DronReal dronReal;
    private String[] usuariosAutorizados = {"admin", "operador1"};

    public ProxyDron(String idDron, String usuario) {
        this.idDron = idDron;
        this.usuario = usuario;
    }

    public void enviarPaquete(String paquete) {
        boolean autorizado = false;
        for (String u : usuariosAutorizados) {
            if (u.equals(usuario)) {
                autorizado = true;
                break;
            }
        }

        if (autorizado) {
            if (dronReal == null) {
                dronReal = new DronReal(idDron);
            }
            dronReal.enviarPaquete(paquete);
        } else {
            System.out.println("Acceso denegado al dron " + idDron + " para el usuario: " + usuario);
        }
    }
}

public class Proxy {
    public static void main(String[] args) {
        Dron dron1 = new ProxyDron("DRN-001", "admin");
        dron1.enviarPaquete("Paquete 1");

        Dron dron2 = new ProxyDron("DRN-002", "invitado");
        dron2.enviarPaquete("Paquete 2");
    }
}
